Update SAML21J18A target to Mbed OS 5.12

janjongboom opened the pull request:
### Description

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->

Co-authored by @martinichka, follow up from https://github.com/ARMmbed/mbed-os/pull/7276.

This PR adds Mbed OS 5.10 (incl. RTOS) support for the SAML21J18A target. It also updates to the latest available device pack from MDK (to Keil.SAM-L_DFP.1.2.1). There are a few issues left to figure out, mostly timer related (see #8656). Tested on: SAML21 XPLAINED PRO development board (already landed in https://github.com/ARMmbed/mbed-ls/pull/390) with GCC ARM 6.

Note that you need the latest interface firmware from the XPLAINED PRO board (automatically pops up when connecting in Atmel Studio).

There are issues with flashing binaries in succession via the USB drag-n-drop interface, which makes the automated testing process very painful (running into SYNC_FAILED often, restarting board solves this). Are there any ways to cut power to a board between GT tests?

#### Todo

* Need to fix #8656.
* Update IAR and ARMCC linker scripts, only verified against GCC ARM.
* Investigate other failing tests.
* Enable LoRaWAN unit tests, not sure how to do that (@AnttiKauppila @hasnainvirk ?) - verified that the device can join the network and send messages, but would be good to test this.

#### Test results

```
mbed-os-features-frameworks-utest-tests-unit_tests-basic_test                               OK
mbed-os-features-frameworks-utest-tests-unit_tests-basic_test_default                       OK
mbed-os-features-frameworks-utest-tests-unit_tests-case_control_async                       OK
mbed-os-features-frameworks-utest-tests-unit_tests-case_control_repeat                      OK
mbed-os-features-frameworks-utest-tests-unit_tests-case_selection                           OK
mbed-os-features-frameworks-utest-tests-unit_tests-case_setup_failure                       OK
mbed-os-features-frameworks-utest-tests-unit_tests-case_teardown_failure                    OK
mbed-os-features-frameworks-utest-tests-unit_tests-minimal_async_scheduler                  OK
mbed-os-features-frameworks-utest-tests-unit_tests-minimal_scheduler                        OK
mbed-os-features-frameworks-utest-tests-unit_tests-test_assertion_failure_test_setup        OK
mbed-os-features-frameworks-utest-tests-unit_tests-test_setup_case_selection_failure        OK
mbed-os-features-frameworks-utest-tests-unit_tests-test_setup_failure                       OK
mbed-os-features-frameworks-utest-tests-unit_tests-test_skip                                OK
mbed-os-features-frameworks-utest-tests-unit_tests-control_type                             OK
mbed-os-features-frameworks-utest-tests-unit_tests-case_async_validate                      FAIL (https://github.com/ARMmbed/mbed-os/issues/8656)
mbed-os-features-storage-tests-blockdevice-buffered_block_device                            OK
mbed-os-features-storage-tests-blockdevice-flashsim_block_device                            OK
mbed-os-features-storage-tests-blockdevice-heap_block_device                                OK
mbed-os-features-storage-tests-blockdevice-mbr_block_device                                 OK
mbed-os-features-storage-tests-blockdevice-util_block_device                                OK
mbed-os-tests-integration-basic                                                             OK
mbed-os-tests-mbed_drivers-c_strings                                                        OK
mbed-os-tests-mbed_drivers-crc                                                              FAIL (Test thread safety)
mbed-os-tests-mbed_drivers-dev_null                                                         OK
mbed-os-tests-mbed_drivers-echo                                                             OK
mbed-os-tests-mbed_drivers-generic_tests                                                    OK
mbed-os-tests-mbed_drivers-sleep_lock                                                       OK
mbed-os-tests-mbed_drivers-stl_features                                                     OK
mbed-os-tests-mbed_functional-callback                                                      OK
mbed-os-tests-mbed_functional-callback_big                                                  OK
mbed-os-tests-mbed_functional-callback_small                                                OK
mbed-os-tests-mbed_functional-functionpointer                                               OK
mbed-os-tests-mbed_hal-critical_section                                                     OK
mbed-os-tests-mbed_hal-rtc                                                                  FAIL (RTC - persist)
mbed-os-tests-mbed_hal-rtc_reset                                                            FAIL
mbed-os-tests-mbed_hal-rtc_time                                                             OK
mbed-os-tests-mbed_hal-rtc_time_conv                                                        OK
mbed-os-tests-mbed_hal-sleep                                                                FAIL (sleep - source of wake-up - us ticker)
mbed-os-tests-mbed_hal-sleep_manager                                                        OK
mbed-os-tests-mbed_hal-ticker                                                               FAIL (test_frequencies_and_masks)
mbed-os-tests-mbed_platform-circularbuffer                                                  OK
mbed-os-tests-mbed_platform-critical_section                                                FAIL (Test critical section C API)
mbed-os-tests-mbed_platform-error_handling                                                  OK
mbed-os-tests-mbed_platform-filehandle                                                      OK
mbed-os-tests-mbed_platform-sharedptr                                                       OK
mbed-os-tests-mbed_platform-singletonptr                                                    OK
mbed-os-tests-mbed_platform-stats_heap                                                      OK
mbed-os-tests-mbed_platform-system_reset                                                    OK
mbed-os-tests-mbed_platform-transaction                                                     OK
mbed-os-tests-mbedmicro-mbed-attributes                                                     OK
mbed-os-tests-mbedmicro-mbed-call_before_main                                               OK
mbed-os-tests-mbedmicro-mbed-cpp                                                            OK
mbed-os-tests-mbedmicro-mbed-div                                                            OK
mbed-os-tests-mbedmicro-mbed-static_assert                                                  OK
mbed-os-tests-mbedmicro-rtos-mbed-kernel_tick_count                                         OK
mbed-os-tests-mbedmicro-rtos-mbed-memorypool                                                OK
mbed-os-tests-mbedtls-multi                                                                 OK
mbed-os-tests-mbedtls-selftest                                                              FAIL (mbedtls_sha512_self_test)
```

[Full test log](https://gist.github.com/janjongboom/e17b3ee7db846b32646d2c471dda08a4)

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [X] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

theotherjimmy reviewed (approved):
No tools changes. Targets.json seems fine to me.

MarceloSalazar commented:
@maclobdell please have a look. I understand you're involved in conversations about SAMx devices.

cmonr commented:
> Are there any ways to cut power to a board between GT tests?

Not that I'm aware of. @ARMmbed/mbed-os-test-team , thoughts?

bridadan commented:
@cmonr Not at the moment no

cmonr commented:
@janjongboom This'll need a rebase to pull the targets.json updates.

@ARMmbed/mbed-os-hal @ashok-rao @ARMmbed/mbed-os-maintainers @maclobdell Please review when able.

cmonr commented:
@janjongboom Mind rebasing any time soon?

@mbed-os-hal @ashok-rao @ARMmbed/mbed-os-maintainers @maclobdell Would anyone else like to review?

janjongboom commented:
@cmonr Yes, will do soon.

janjongboom commented:
@cmonr Rebased against master, verified blinky against RTOS and non-RTOS builds. Here are the test results:

```
mbed-os-features-frameworks-utest-tests-unit_tests-basic_test                   OK
mbed-os-features-frameworks-utest-tests-unit_tests-basic_test_default           OK
mbed-os-features-frameworks-utest-tests-unit_tests-case_async_validate          FAILED - Validate: Simple Validation
mbed-os-features-frameworks-utest-tests-unit_tests-case_control_async           OK
mbed-os-features-frameworks-utest-tests-unit_tests-case_control_repeat          OK
mbed-os-features-frameworks-utest-tests-unit_tests-case_selection               OK
mbed-os-features-frameworks-utest-tests-unit_tests-case_setup_failure           OK
mbed-os-features-frameworks-utest-tests-unit_tests-case_teardown_failure        OK
mbed-os-features-frameworks-utest-tests-unit_tests-control_type                 OK
mbed-os-features-frameworks-utest-tests-unit_tests-minimal_async_scheduler      OK
mbed-os-features-frameworks-utest-tests-unit_tests-minimal_scheduler            OK
mbed-os-features-frameworks-utest-tests-unit_tests-test_assertion_failure_test_setup    OK
mbed-os-features-frameworks-utest-tests-unit_tests-test_setup_case_selection_failure    OK
mbed-os-features-frameworks-utest-tests-unit_tests-test_setup_failure           OK
mbed-os-features-frameworks-utest-tests-unit_tests-test_skip                    OK
mbed-os-features-storage-tests-blockdevice-buffered_block_device                OK
mbed-os-features-storage-tests-blockdevice-flashsim_block_device                OK
mbed-os-features-storage-tests-blockdevice-general_block_device                 OK
mbed-os-features-storage-tests-blockdevice-heap_block_device                    OK (but I think I saw a hard fault at one point too)
mbed-os-features-storage-tests-blockdevice-mbr_block_device                     OK
mbed-os-features-storage-tests-blockdevice-util_block_device                    OK
mbed-os-features-storage-tests-kvstore-filesystemstore_tests                    OK
mbed-os-features-storage-tests-kvstore-general_tests_phase_1                    OK
mbed-os-features-storage-tests-kvstore-general_tests_phase_2                    OK
mbed-os-features-storage-tests-kvstore-static_tests                             OK
mbed-os-features-storage-tests-kvstore-tdbstore_whitebox                        OK
mbed-os-tests-integration-basic                                                 OK
mbed-os-tests-mbed_drivers-c_strings                                            OK
mbed-os-tests-mbed_drivers-crc                                                  OK
mbed-os-tests-mbed_drivers-dev_null                                             OK
mbed-os-tests-mbed_drivers-echo                                                 OK
mbed-os-tests-mbed_drivers-generic_tests                                        OK
mbed-os-tests-mbed_drivers-sleep_lock                                           OK
mbed-os-tests-mbed_drivers-stl_features                                         OK
mbed-os-tests-mbed_functional-callback                                          OK
mbed-os-tests-mbed_functional-callback_big                                      OK
mbed-os-tests-mbed_functional-callback_small                                    OK
mbed-os-tests-mbed_functional-functionpointer                                   OK
mbed-os-tests-mbed_hal-critical_section                                         OK
mbed-os-tests-mbed_hal-minimum_requirements                                     OK
mbed-os-tests-mbed_hal-rtc                                                      FAILED - RTC - persist
mbed-os-tests-mbed_hal-rtc_reset                                                FAILED - mbed-os-tests-mbed_hal-rtc_reset
mbed-os-tests-mbed_hal-rtc_time                                                 OK
mbed-os-tests-mbed_hal-rtc_time_conv                                            OK
mbed-os-tests-mbed_hal-sleep                                                    FAILED - sleep - source of wake-up - us ticker
mbed-os-tests-mbed_hal-sleep_manager                                            OK
mbed-os-tests-mbed_hal-ticker                                                   FAILED - test_frequencies_and_masks
mbed-os-tests-mbed_platform-circularbuffer                                      OK
mbed-os-tests-mbed_platform-critical_section                                    FAILED - Test critical section C API
mbed-os-tests-mbed_platform-error_handling                                      OK
mbed-os-tests-mbed_platform-filehandle                                          OK
mbed-os-tests-mbed_platform-sharedptr                                           OK
mbed-os-tests-mbed_platform-singletonptr                                        OK
mbed-os-tests-mbed_platform-stats_heap                                          OK
mbed-os-tests-mbed_platform-stream                                              OK
mbed-os-tests-mbed_platform-system_reset                                        OK
mbed-os-tests-mbed_platform-transaction                                         OK
mbed-os-tests-mbedmicro-mbed-attributes                                         OK
mbed-os-tests-mbedmicro-mbed-call_before_main                                   OK
mbed-os-tests-mbedmicro-mbed-cpp                                                OK
mbed-os-tests-mbedmicro-mbed-div                                                OK
mbed-os-tests-mbedmicro-mbed-static_assert                                      OK
mbed-os-tests-mbedmicro-rtos-mbed-kernel_tick_count                             OK
mbed-os-tests-mbedmicro-rtos-mbed-memorypool                                    OK (but I think I saw a hard fault after all the tests ran)
mbed-os-tests-mbedtls-multi                                                     OK
mbed-os-tests-mbedtls-selftest                                                  FAILED - mbedtls_sha512_self_test, but I think it's a timeout on a slow processor
```

I think some of these are related to the us ticker not working if set to <1 ms (#8656). But this is an issue already present in the current target.

I could change the target so that it sets the minimum ticker time to 1000 us. At least the behavior would be better than it is now (the ticker never fires). @cmonr What do you think about that?

cmonr commented:
> I could change the target so that it sets the minimum ticker time to 1000 us. At least the behavior would be better than it is now (the ticker never fires). @cmonr What do you think about that?

Go ahead and make the change for now. It won't be the worst thing, especially since there's already an issue open for it.

janjongboom commented:
@cmonr OK done. I've tracked it down to some register not set correctly when triggering the IRQ manually, but I have no clue about Atmel IRQ handling, and how to solve this. I've put a big comment block in the SAML21 HAL driver.

The `mbed-os-features-frameworks-utest-tests-unit_tests-case_async_validate` test now passes, but the rest still fails.

janjongboom commented:
Pushed the FlashIAP implementation as well.

All flash tests pass (`mbed-os-tests-mbed_drivers-flashiap`, `mbed-os-components-storage-blockdevice-component_flashiap-tests-filesystem-fopen`, `mbed-os-features-storage-tests-blockdevice-flashsim_block_device` and `mbed-os-tests-mbed_hal-flash`) except:

1. `flash_clock_and_cache_test` in the functional tests fail, because `time_cpu_cycles` times out. Maybe related to other test failures? At least that test doesn't do anything with the FlashIAP interface.
2. Need to change `MAX_TEST_SIZE` in the storage/blockdevice/FlashIAP tests to 64K, and reducing the size of files in `fslittle_fopen_test_15`. The target does not have so much free flash.

janjongboom commented on targets/TARGET_Atmel/TARGET_SAM_CortexM0P/us_ticker.c:
```diff
@@ -127,37 +129,33 @@ uint32_t us_ticker_read()
-    uint32_t cur_time;
-    int32_t delta;
-
-    cur_time = us_ticker_read();
```
This is already done by the Mbed ticker implementation.

janjongboom commented:
@cmonr How can we progress this? I'd like to have this mainlined before 25 January.

This PR brings the target from unusable to usable with some tests failing, and a good basis to fix the remaining issues.

0xc0170 commented:
As the target is not supported anymore, can still stay so no version label for this PR. 
Fixes would land only on master and wait for other fixes as stated above.

janjongboom commented:
@0xc0170 What do you mean with 'no version label'? Fix on master and thus inclusion in the 5.12 release is fine for me.

0xc0170 commented:
> @0xc0170 What do you mean with 'no version label'? Fix on master and thus inclusion in the 5.12 release is fine for me.

That is what I meant, sorry for the confusion.

janjongboom commented:
@0xc0170 Check, that works for me.

cmonr commented:
@mbed-os-hal @ashok-rao @maclobdell Please review.

janjongboom commented:
Any update here?

0xc0170 commented on targets/targets.json:
```diff
@@ -6384,6 +6389,7 @@
             "SPISLAVE",
             "SPI_ASYNCH"
         ],
+        "release_versions": ["2", "5"],
```
As this is not yet completed, do we keep release_versions ?

janjongboom replied:
I'm not sure how this is supposed to work, @cmonr ? I thought we need this to get it to compile with Mbed OS at all?

cmonr replied:
Yup, it's needed to test.

I think only `5` is needed if non-rtos is still not working.

janjongboom replied:
@cmonr, Looking through targets.json it seems most targets have both 2 and 5. I think it's required to keep new versions of Mbed 2 working with the SAML21 target. I know at least one company which uses this.

0xc0170 commented on components/storage/blockdevice/COMPONENT_FLASHIAP/TESTS/filesystem/fopen/fopen.cpp:
```diff
@@ -78,7 +78,11 @@ static char fslittle_fopen_utest_msg_g[FSLITTLE_UTEST_MSG_BUF_SIZE];
 #define FSLITTLE_FOPEN_TEST_WORK_BUF_SIZE_1    64
 #define FSLITTLE_FOPEN_TEST_FILEPATH_MAX_DEPTH 20
+#ifdef TARGET_SAML21J18A
+static const int MAX_TEST_SIZE = 64 * 1024;
```
test should not contain specific target. Wasn't this already addressed on master? 

cc @ARMmbed/mbed-os-storage

janjongboom replied:
@0xc0170 What do you suggest in that case? The max size is now 512K which is way too much. The target only has 256K in total. I assume this will be an issue in many other targets too.

davidsaada replied:
@janjongboom Sorry for missing this earlier. Your target is not the only one with small RAM. I guess you're trying to run all the tests and see if they pass on your target. However, in the LittleFS specific case, tests are only limited in our CI to one or two boards that are capable of running the tests (have enough memory, have persistent storage etc.). So as @0xc0170 said - no target specific ifdefs should appear on the test code.

janjongboom replied:
@davidsaada But my target can run the tests, just the size should be limited. I think it's a nice test to have actually to verify that FlashIAP works properly.

davidsaada replied:
Nevertheless, adding specific target based ifdefs in our test code is not the way to go. Other targets may have less memory (there are such I know), meaning that supporting them would require additional ifdefs. There are dozens of supported targets and choosing this path will make the code unreadable. We can try considering the available memory to determine the test size in this test, but it should be outside this PR's scope IMO.

0xc0170 replied:
Can we remove this from this PR, create an issue and address this ? If this is causing an issue for this target, would lead to disabling flashiap until the fix is in place - it would not block the further progress.

janjongboom replied:
@0xc0170 @davidsaada @cmonr I've moved this to mbed_lib, just like the other flashiap configuration (like where the flash starts). Does this work for you? This removes any ifdef code and the default behavior stays the same.

0xc0170 commented on targets/TARGET_Atmel/TARGET_SAM_CortexM0P/TARGET_SAML21J18A/device/TOOLCHAIN_GCC_ARM/saml21j18a.ld:
```diff
@@ -2,26 +2,42 @@ OUTPUT_FORMAT("elf32-littlearm", "elf32-littlearm", "elf32-littlearm")
 
-    /* The stack size used by the application. NOTE: you need to adjust according to your application. */
-    STACK_SIZE = DEFINED(STACK_SIZE) ? STACK_SIZE : DEFINED(__stack_size__) ? __stack_size__ : 0x2000;
+STACK_SIZE = DEFINED(STACK_SIZE) ? STACK_SIZE : DEFINED(__stack_size__) ? __stack_size__ : 4096;
```
Does this need to be updated based on 5.12 stack unification (already on master) ?

janjongboom replied:
Yes, updated. Thanks! Worked out of the box.

0xc0170 commented on targets/TARGET_Atmel/TARGET_SAM_CortexM0P/TARGET_SAML21J18A/device/TOOLCHAIN_GCC_ARM/startup_saml21.c:
```diff
@@ -258,14 +267,17 @@ void Reset_Handler(void)
+#ifdef MBED_CONF_RTOS_PRESENT
+    software_init_hook();
+#else
+    __libc_init_array();
```
Doesn't mbed_boot code file already provide these for Mbed OS 5 ? See mbed_boot.c file, `mbed_start` function.

See the comments in the file:

> Reset vector is
 * responsible for low level platform init and then calling in libc (__main). Depending on compiler and version of C
 * library, predefined function will be called which is implemented by mbed OS.

janjongboom replied:
@0xc0170 How do I set the reset vector to behave like this then? Because I copied this behavior from some NXP targets ([example](https://github.com/ARMmbed/mbed-os/blob/master/targets/TARGET_NXP/TARGET_LPC11U6X/device/TOOLCHAIN_GCC_ARM/TARGET_LPC11U68/startup_LPC11U68.cpp#L152)) and it seems to properly boot in both RTOS and non-RTOS mode. I don't really get how other targets do it because it's all in .S files not in C. Also this code is in the reset handler already, so what's the difference?

0xc0170 replied:
Understood, looks fine

0xc0170 commented on targets/TARGET_Atmel/TARGET_SAM_CortexM0P/flash_api.c:
```diff
@@ -0,0 +1,172 @@
+#include "flash_api.h"
```
license header missing , please add

janjongboom replied:
Done.

0xc0170 commented on targets/TARGET_Atmel/TARGET_SAM_CortexM0P/flash_api.c:
```diff
@@ -0,0 +1,172 @@
+ * @param obj The flash object
+ * @return The flash region size
+ */
+uint32_t flash_get_size(const flash_t *obj) {
```
other HAL files follow mostly our coding style, `{` new line for functions in this file

```
uint32_t flash_get_size(const flash_t *obj)
{

}

```

janjongboom replied:
Done.

janjongboom commented:
@cmonr @0xc0170 I have rebased against master and updated based on @0xc0170's comments. I'd be happy to fix the boot code if someone has an example of how this should be done properly. Please let me know how we can progress this further.

0xc0170 reviewed (approved):
Waiting for @cmonr regarding the release version if its OK as it is, and flashiap test modification, the rest looks fine to me.

0xc0170 reviewed (changes requested):
Waiting for @cmonr regarding the release version if its OK as it is, and flashiap test modification, the rest looks fine to me.

(my previous one was supposed to be request changes)

We can run CI  and have this tested after questions are answered

janjongboom commented:
@0xc0170 @cmonr Updated based on your comments, please let me know.

cmonr commented:
CI started

0xc0170 reviewed (approved):
@ARMmbed/mbed-os-storage  to approve the configuration

0xc0170 commented:
Testing reported an error (pr-head): `tools.utils.InvalidReleaseTargetException: Target 'SAML21J18A' must support ALL of the folowing toolchains to be included in the mbed OS 5.0 official release: ARM, GCC_ARM, IAR`

dannybenor commented:
@offirko can you please review the changes in components/storage/blockdevice/COMPONENT_FLASHIAP

janjongboom commented:
@0xc0170 There is a startup script in the IAR folder (although it doesn't seem to do much), but no linker script. I have no idea how to write this either, and I don't have IAR. But if I remove '5' from the target version list then I cannot compile any programs anymore (`Could not compile for SAML21J18A: Target does not support mbed OS 5`). Could we find a solution for this on short notice?

janjongboom commented:
@cmonr @0xc0170 Can we land this in a feature branch until the IAR linker script is written? We do an announcement about this next week and I don't want to send people to my GitHub page.

Commit:
Swapping I2C pins between original and extended PAD

Looking at the board layout (SAML21 Xplained Pro) we can see that out of
three EXT pinouts only two (EXT1 and EXT3) have a set of four pins
available on a single SERCOM. For example (PB16, PB17, PB22, PB23) on
EXT3 are mapped to SERCOM5. So if we wish to add a peripheral that uses
4 pin SPI like a SX1276 Wing radio, we can only choose EXt3 or EXT1.

However, we found out that we can't use radio shield on EXT3 as the pin
PB10 used as dio2 interrupt line is also mapped to the LED.
For both SPI and I2C peripherlas, HAL chooses Default PAD. This would
result in picking up SERCOM0 for I2C pins (PA08 and PA09) which will not
work for us as we have attached the secure element to EXT3 and we need
to map these pins to SERCOM2.

One way of solving the issue is to mark PA08 and PA09 as NC on SERCOM0
by default. Another is to swap the SERCOMx mapping for the PA08 and PA09
between the Default PAD and Extended PAD.

A proper fix would probably be to physically cut PB10 and LED
connection. This would mean that we don't need to hack the pin map.

Commit:
Adding us ticker to SAML21

0xc0170 commented:
> @cmonr @0xc0170 Can we land this in a feature branch until the IAR linker script is written? We do an announcement about this next week and I don't want to send people to my GitHub page.

We can create a feature branch, but it still should pass tests (in this case without IAR support, it would fail?).


Here's some good example for the icf (I checked one Sam device and adjusted sections):

```
if (!isdefinedsymbol(MBED_APP_START)) { define symbol MBED_APP_START = 0x00000000; }
if (!isdefinedsymbol(MBED_APP_SIZE)) { define symbol MBED_APP_SIZE = 0x80000; }
/*###ICF### Section handled by ICF editor, don't touch! ****/
/*-Editor annotation file-*/
/* IcfEditorFile="$TOOLKIT_DIR$\config\ide\IcfEditor\cortex_v1_0.xml" */
/*-Specials-*/
define symbol __ICFEDIT_intvec_start__ = MBED_APP_START;
/*-Memory Regions-*/
define symbol __ICFEDIT_region_ROM_start__   = MBED_APP_START;
define symbol __ICFEDIT_region_ROM_end__     = MBED_APP_START + MBED_APP_SIZE - 1;
define symbol __ICFEDIT_region_NVIC_start__   = 0x20000000;
define symbol __ICFEDIT_region_NVIC_end__   = 0x200000B8;
define symbol __ICFEDIT_region_RAM_start__   = 0x200000B8;
define symbol __ICFEDIT_region_RAM_end__     = 0x20008000;

/*-Sizes-*/
if (!isdefinedsymbol(MBED_BOOT_STACK_SIZE)) {
    define symbol MBED_BOOT_STACK_SIZE = 0x400;
}

define symbol __ICFEDIT_size_cstack__   = MBED_BOOT_STACK_SIZE;
define symbol __ICFEDIT_size_heap__     = 0x400;
/**** End of ICF editor section. ###ICF###*/

define memory mem with size = 4G;
define region ROM_region   = mem:[from __ICFEDIT_region_ROM_start__   to __ICFEDIT_region_ROM_end__];
define region RAM_region   = mem:[from __ICFEDIT_region_RAM_start__   to __ICFEDIT_region_RAM_end__];

define block CSTACK    with alignment = 8, size = __ICFEDIT_size_cstack__   { };
define block HEAP      with alignment = 8, size = __ICFEDIT_size_heap__     { };


initialize by copy { readwrite };
do not initialize  { section .noinit };

place at address mem:__ICFEDIT_intvec_start__ { readonly section .intvec };
place in ROM_region   { readonly };
place in RAM_region   { readwrite,
                        block HEAP, block CSTACK };
```

Just set RAM, NVIC regions and heap size, this should help you. I assumed there's only one RAM region at address 0x20000000.

janjongboom commented:
@0xc0170 I assume this still requires a startup script right? It seems empty for this target right now.

0xc0170 commented:
>@0xc0170 I assume this still requires a startup script right? It seems empty for this target right now.

Yes, what about feature branch? Or is this targeting master (needs IAR support still) ?

c1728p9 commented on targets/TARGET_Atmel/TARGET_SAM_CortexM0P/utils/cmsis/TARGET_SAML21/source/system_saml21.c:
```diff
@@ -50,7 +50,7 @@
  *  the source for the main clock at chip startup.
  */
-#define __SYSTEM_CLOCK    (4000000)
+#define __SYSTEM_CLOCK    (4000000 * 12)
```
The comment indicates that this is the "System RC Oscillator" which is typically internal to the chip. Why is the being multiplied by 12? If a clock needs to be updated shouldn't it be for an external or PLL clock?

janjongboom replied:
Good question, the previous version ran on 4 MHz. In this PR we're upping the clock frequency to 48 MHz (normal speed of the device). However, if this define is not in sync with the actual clock speed all timers are off in Mbed OS. But I cannot update this value after boot (I've tried but it doesn't do anything), so I think it should be the value after boot here, not the one during boot. Have to update the comment though.

c1728p9 commented on targets/TARGET_Atmel/TARGET_SAM_CortexM0P/drivers/system/clock/TARGET_SAML21/clock_feature.h:
```diff
@@ -948,9 +948,9 @@ static inline void system_main_clock_set_failure_detect(
 {
     if (enable) {
-        MCLK->CTRLA.reg |=  MCLK_CTRLA_CFDEN;
+        MCLK->CTRLA.reg |=  4;
```
Is there a define that can be used here rather than 4?

c1728p9 commented on components/storage/blockdevice/COMPONENT_FLASHIAP/mbed_lib.json:
```diff
@@ -8,12 +8,26 @@
+            "base-address": "0x30000",
+            "size": "0x10000",
+            "num-blocks": 3,
+            "max-test-size": "64 * 1024"
```
Why does the size need to be reduced for this target? Not enough storage space?

janjongboom replied:
Yes. Only 256K flash and half is already taken by the binary, so too little to use bigger FlashIAP file system.

c1728p9 reviewed (approved):
A few minor comments, but otherwise looks good to me.

janjongboom committed:
Update SAML21J18A target to Mbed OS 5.12

janjongboom committed:
Fix compilation on SAML21J18A

janjongboom committed:
Fix issue with prescaler
